src/services/codefixes/convertTypedefToType.ts

```diff
@@ -0,0 +1,253 @@
+    }
+}
+
+function createInterfaceForTypeLiteral(
+    typeName: string,
+    typeLiteral: JSDocTypeLiteral
+): InterfaceDeclaration | undefined {
+    const propertySignatures = createSignatureFromTypeLiteral(typeLiteral);
+    if (!some(propertySignatures)) return;
+    const interfaceDeclaration = factory.createInterfaceDeclaration(
+        [],
+        typeName,
+        [],
+        [],
+        propertySignatures,
+    );
+    return interfaceDeclaration;
+}
+
+function createTypeAliasForTypeExpression(
+    typeName: string,
+    typeExpression: JSDocTypeExpression
+): TypeAliasDeclaration | undefined {
+    const typeReference = createTypeReference(typeExpression.type);
+    if (!typeReference) return;
+    const declaration = factory.createTypeAliasDeclaration(
+        [],
+        factory.createIdentifier(typeName),
+        [],
+        typeReference
+    );
+    return declaration;
+}
+
+function createSignatureFromTypeLiteral(typeLiteral: JSDocTypeLiteral): PropertySignature[] | undefined {
+    const propertyTags = typeLiteral.jsDocPropertyTags;
+    if (!some(propertyTags)) return;
+
+    const getSignature = (tag: JSDocPropertyTag) => {
+        const name = getPropertyName(tag);
+        const type = tag.typeExpression?.type;
+        const isOptional = tag.isBracketed;
+        let typeReference;
+
+        // Recursively handle nested object type
+        if (type && type.kind === SyntaxKind.JSDocTypeLiteral) {
+            const signatures = createSignatureFromTypeLiteral(type as JSDocTypeLiteral);
+            typeReference = factory.createTypeLiteralNode(signatures);
+        }
+        // Leaf node, where type.kind === SyntaxKind.JSDocTypeExpression
+        else if (type) {
+            typeReference = createTypeReference(type);
+        }
+
+        if (typeReference && name) {
+            const questionToken = isOptional ? factory.createToken(SyntaxKind.QuestionToken) : undefined;
+            const prop = factory.createPropertySignature(
+                [],
+                name,
+                questionToken,
+                typeReference
+            );
+
+            return prop;
+        }
+    };
+
+    const props = mapDefined(propertyTags, getSignature);
+    return props;
+}
+
+function getPropertyName(tag: JSDocPropertyTag): string | undefined {
+    const { name } = tag;
+    if (!name) return;
+
+    let propertyName;
+    // for "@property {string} parent.child" or "@property {string} parent.child.grandchild" in nested object type
+    // We'll get "child" in the first example or "grandchild" in the second example as the prop name
+    if (name.kind === SyntaxKind.QualifiedName) {
+        propertyName = name.right.getFullText().trim();
+    }
+    else {
+        propertyName = tag.name.getFullText().trim();
+    }
+    return propertyName;
+}
+
+// Create TypeReferenceNode when we reach the leaf node of AST
+function createTypeReference(type: TypeNode): TypeNode | undefined {

```

Comment:
This function seems leave out more than it covers (intersections, object types, type references, conditional types, etc.). Why can’t we just clone the type node that we got from the JSDoc type expression?

Reply:
done